modules/navier_stokes/include/loops/GatherRCDataFaceThread.h

```diff
@@ -0,0 +1,212 @@
+};
+
+template <typename RangeType>
+GatherRCDataFaceThread<RangeType>::GatherRCDataFaceThread(FEProblemBase & fe_problem,
+                                                          const std::vector<unsigned int> & vars)
+  : ThreadedFaceLoop<RangeType>(fe_problem, {}), _vars(vars)
+{
+}
+
+template <typename RangeType>
+GatherRCDataFaceThread<RangeType>::GatherRCDataFaceThread(GatherRCDataFaceThread & x,
+                                                          Threads::split split)
+  : ThreadedFaceLoop<RangeType>(x, split), _vars(x._vars)
+{
+}
+
+template <typename RangeType>
+template <typename... Attribs>
+void
+GatherRCDataFaceThread<RangeType>::getVarROs(std::vector<INSFVMomentumResidualObject *> & ros,
+                                             TheWarehouse::QueryCache<Attribs...> & queries)
+{
+  for (const auto var_num : _vars)
+  {
+    // We don't want to do cascading var num attributes or else the second time around we won't get
+    // any results out of the query (e.g. an object cannot have a variable that simultaneously has
+    // both var number 0 and 1)
+    auto copied_queries = queries;
+    std::vector<INSFVMomentumResidualObject *> var_ros;
+    copied_queries.template condition<AttribVar>(static_cast<int>(var_num)).queryInto(var_ros);
+    for (auto * const var_ro : var_ros)
+      ros.push_back(var_ro);
+  }
+}
+
+template <typename RangeType>
+void
+GatherRCDataFaceThread<RangeType>::onFace(const FaceInfo & fi)
+{
+  for (auto * const k : _fv_flux_kernels)
+    k->gatherRCData(fi);
+}
+
+template <typename RangeType>
+void
+GatherRCDataFaceThread<RangeType>::onBoundary(const FaceInfo & fi, BoundaryID bnd_id)
+{
+  {
+    std::vector<INSFVMomentumResidualObject *> bcs;
+    // cannot bind to lvalue reference otherwise the temporary is destroyed and we are left with a
+    // dangling reference
+    auto queries = this->_fe_problem.theWarehouse()
+                       .query()
+                       .template condition<AttribSystem>("FVFluxBC")
+                       .template condition<AttribThread>(this->_tid)
+                       .template condition<AttribBoundaries>(bnd_id);
+    getVarROs(bcs, queries);
+
+    for (auto * const bc : bcs)
+      bc->gatherRCData(fi);
+  }
+
+  {
+    std::vector<INSFVMomentumResidualObject *> iks;
+    // cannot bind to lvalue reference otherwise the temporary is destroyed and we are left with a
+    // dangling reference
+    auto queries = this->_fe_problem.theWarehouse()
+                       .query()
+                       .template condition<AttribSystem>("FVInterfaceKernel")
+                       .template condition<AttribThread>(this->_tid)
+                       .template condition<AttribBoundaries>(bnd_id);
+    getVarROs(iks, queries);
+
+    for (auto * const ik : iks)
+      ik->gatherRCData(fi);
+  }
+}
+
+template <typename RangeType>
+void
+GatherRCDataFaceThread<RangeType>::subdomainChanged()
+{
+  ThreadedFaceLoop<RangeType>::subdomainChanged();
```

Comment:
can boundaries and interfaces be changed on a subdomain changed?

Reply:
There's not really a connection between subdomains and those objects. You could be on the same subdomain and have a different set of bcs and iks just depending on what side of an element you are on. We re-compute the current set of bcs and iks with every call to `onBoundary`